lgc/patch/PatchReadFirstLane.cpp

```diff
@@ -0,0 +1,418 @@
+
+  for (Instruction &inst : llvm::reverse(*block)) {
+    // Skip readfirstlane-irrelevant instructions or a source of divergence which isn't clear to determine its operands
+    if (m_canAssumeUniformDivergentUsesMap.count(&inst) == 0 || (m_targetTransformInfo->isSourceOfDivergence(&inst)))
+      continue;
+
+    // For all operands, see if they can be assumed uniform
+    for (Use &use : inst.operands()) {
+      if (!m_divergenceAnalysis->isDivergentUse(&use))
+        continue;
+      auto operandInst = dyn_cast<Instruction>(use.get());
+      if (!operandInst)
+        continue;
+      if (m_canAssumeUniformDivergentUsesMap.count(operandInst) == 1)
+        continue;
+      if (operandInst->getParent() != block || !isAllUsersDefinedInBlock(operandInst, block))
+        continue;
+      // The operand can be assumed uniform becasue all users are assumed uniform
+      if (isAllUsersAssumedUniform(operandInst)) {
+        m_canAssumeUniformDivergentUsesMap[operandInst] = {};
+        m_canAssumeUniforms.push_back(operandInst);
+      }
+    }
+
+    bool canCollectOperands = true;
+    // NOTE: Stop propagation for a loadInst whose address space is fat pointer because it has uncertain size
+    auto loadInst = dyn_cast<LoadInst>(&inst);
+
+    // For all operands, see if all operands can be assumed uniform and are in the same block. If true, we can collect
+    // the operands to be used for the propagation heuristic in findInsertLocation.
+    SmallVector<Instruction *, 2> operandInsts;
+    for (Use &use : inst.operands()) {
+      if (auto operandInst = dyn_cast<Instruction>(use.get())) {
+        if (!m_canAssumeUniformDivergentUsesMap.count(operandInst))
+          continue;
+        if (operandInst->getParent() != block || !isAllUsersDefinedInBlock(operandInst, block) ||
+            (loadInst && loadInst->getPointerAddressSpace() == ADDR_SPACE_BUFFER_FAT_POINTER)) {
+          canCollectOperands = false;
+          break;
+        }
+        operandInsts.push_back(operandInst);
+      }
+    }
+    if (canCollectOperands)
+      m_canAssumeUniformDivergentUsesMap[&inst] = std::move(operandInsts);
+  }
+}
+
+// =====================================================================================================================
+// Find the best insert locations according to the heuristic.
+// The heuristic is: if an instruction that can be assumed to be uniform has multiple divergent operands, then you take
+// the definition of the divergent operand that is earliest in basic block order (call it "the earliest divergent
+// operand") and propagate up to that instruction; if it turns out that that instruction can be assumed to be uniform,
+// then we can just insert the readfirstlane there (or propagate).
+//
+// @param block : The process basic block
+// @param readFirstLaneCount : The initial amdgcn_readfirstlane vector
+void PatchReadFirstLane::findInsertLocations(BasicBlock *block,
```

Comment:
I would like to propose some implementation I came up which sounds more easy to understand.
btw. I think the function is better to be renamed as `findBestInsertLocation()`.
I just simply paste my implementation here, @nhaehnle @xuechen417 please help take a look whether it sounds good. I have not tested yet.

```
// We want instructions appear late in the IR put early in the set. That is to say the instructions are stored in the
// reverse order how they appear in the IR.
static bool instructionOrdering(const Instruction *p1, const Instruction *p2) {
  if (p1->comesBefore(p2))
    return false;
  return true;
}

void PatchReadFirstLane::findInsertLocations(BasicBlock *block,
                                             const SmallVectorImpl<Instruction *> &initialReadFirstLanes) {
  m_insertLocations.clear();

  // As we are trying to lift readfirstlane(i.e. RFL) instruction up in the IR. we must ensure we will not
  // introduce any extra RFLs. If one instruction access more than one divergent sources, we will do some lookahead
  // search here, to see whether these divergent sources come from the same value (i.e. they have a common predecessor
  // in def-use graph) so we can still move the RFL up in the IR to the common predecessor. If we cannot prove
  // that, just insert after the current instruction. This can ensure we are not inserting extra RFL during the lift
  // process.
  DenseSet<Instruction *> visited;
  for (auto &initialReadFirstLane : initialReadFirstLanes) {
    // Find an insert location for each initial readfirstlane
    std::queue<Instruction *> candidateQ;
    candidateQ.push(initialReadFirstLane);
    while (!candidateQ.empty()) {
      auto candidate = candidateQ.front();
      candidateQ.pop();

      if (visited.count(candidate))
        continue;

      visited.insert(candidate);

      auto &sources = m_canAssumeUniformDivergentUsesMap[candidate];
      unsigned divergentSrcNum = sources.size();
      if (divergentSrcNum == 0) {
        // The candidate has no divergent source, we cannot lift RFL further.
        addLocation(candidate);
      } else if (divergentSrcNum == 1) {
        candidateQ.push(sources[0]);
      } else {
#define MAX_SEARCH_STEP 10
        // we have more than one divergent sources, check whether we can prove they come from the same divergent value.
        std::set<Instruction *, decltype(&instructionOrdering)> divergentOrigins;
        for (auto s : sources)
          divergentOrigins.insert(s);

        unsigned searchStep = 0;
        while (divergentOrigins.size() > 1) {
          auto bottom = divergentOrigins.begin();

          auto &sources2 = m_canAssumeUniformDivergentUsesMap[*bottom];
          // TODO I didn't think further about this, stop for now for simplicity.
          if (!sources2.size())
            break;

          divergentOrigins.erase(bottom);

          for (auto s : sources2)
            divergentOrigins.insert(s);

          if (searchStep++ > MAX_SEARCH_STEP)
            break;
        }
        // If we successfully prove that the divergent sources of the current instruction shares the same origin,
        // Let's continue search, else let's stop at current candidate.
        if (divergentOrigins.size() == 1)
          candidateQ.push(*divergentOrigins.begin());
        else
          addLocation(candidate);
      }
    }
  }
}
```

Reply:
I have re-work the logic of this function which is more general than the last version. Please review.